alicloud/resource_alicloud_router_interface.go

```diff
@@ -0,0 +1,321 @@
+	conn := meta.(*AliyunClient).ecsconn
+
+	filter := ecs.Filter{Key: "RouterInterfaceId", Value: []string{d.Id()}}
+	args := &ecs.DescribeRouterInterfacesArgs{
+		RegionId: common.Region(d.Get("region").(string)),
+		Filter:   []ecs.Filter{filter},
+	}
+	resp, err := conn.DescribeRouterInterfaces(args)
+	if err != nil {
+		return fmt.Errorf("DescribeRouterInterfaces got an error: %#v", err)
+	}
+
+	routerInterface := resp.RouterInterfaceSet.RouterInterfaceType
+	if len(routerInterface) == 0 {
+		return fmt.Errorf("No router interface found.")
+	}
+
+	for _, ri := range routerInterface {
+		if ri.RouterInterfaceId == d.Id() {
+			d.Set("role", ri.Role)
+			d.Set("specification", ri.Spec)
+			d.Set("name", ri.Name)
+			d.Set("router_id", ri.RouterId)
+			d.Set("router_type", ri.RouterType)
+			d.Set("description", ri.Description)
+			d.Set("access_point_id", ri.AccessPointId)
+			d.Set("opposite_access_point_id", ri.OppositeAccessPointId)
+			d.Set("opposite_router_type", ri.OppositeRouterType)
+			d.Set("opposite_router_id", ri.OppositeRouterId)
+			d.Set("opposite_interface_id", ri.OppositeInterfaceId)
+			d.Set("opposite_interface_owner_id", ri.OppositeInterfaceOwnerId)
+			d.Set("health_check_source_ip", ri.HealthCheckSourceIp)
+			d.Set("health_check_target_ip", ri.HealthCheckTargetIp)
+			return nil
+		}
+	}
+
+	d.SetId("")
+	return fmt.Errorf("No router interface found.")
+}
+
+func resourceAlicloudRouterInterfaceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).ecsconn
+
+	args := &ecs.OperateRouterInterfaceArgs{
+		RegionId:          common.Region(d.Get("region").(string)),
+		RouterInterfaceId: d.Id(),
+	}
+
+	return resource.Retry(5*time.Minute, func() *resource.RetryError {
+		if _, err := conn.DeleteRouterInterface(args); err != nil {
+			if IsExceptedError(err, RouterInterfaceIncorrectStatus) || IsExceptedError(err, DependencyViolationRouterInterfaceReferedByRouteEntry) {
+				return resource.RetryableError(fmt.Errorf("Router interface in use - trying again while it is deleted."))
+			}
+			return resource.NonRetryableError(fmt.Errorf("Error deleting interface %s: %#v", d.Id(), err))
+		}
+		return nil
+	})
+}
+
+func buildAlicloudRouterInterfaceCreateArgs(d *schema.ResourceData, meta interface{}) (*ecs.CreateRouterInterfaceArgs, error) {
+	args := &ecs.CreateRouterInterfaceArgs{
+		RegionId:           common.Region(d.Get("region").(string)),
+		RouterType:         ecs.RouterType(d.Get("router_type").(string)),
+		RouterId:           d.Get("router_id").(string),
+		Role:               ecs.Role(d.Get("role").(string)),
+		Spec:               ecs.Spec(d.Get("specification").(string)),
+		OppositeRegionId:   common.Region(d.Get("opposite_region").(string)),
+		OppositeRouterType: ecs.RouterType(d.Get("opposite_router_type").(string)),
+	}
+
+	_, sourceOk := d.GetOk("health_check_source_ip")
+	_, targetOk := d.GetOk("health_check_target_ip")
+	if sourceOk && !targetOk || !sourceOk && targetOk {
```

Comment:
Put this check before defining args would be better. In addition, do we need to check their values are ""?